testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,242 @@
+	limit.rlim_cur = limit.rlim_max;
+	SAFE_SETRLIMIT(RLIMIT_CORE, &limit);
+
+	switch (prctl(PR_GET_DUMPABLE)) {
+	case 0:
+		tst_brk(TCONF, "Process is not dumpable.");
+	case 1:
+		break;
+	default:
+		tst_brk(TBROK | TERRNO, "prctl(PR_GET_DUMPABLE)");
+	}
+
+	SAFE_FILE_SCANF(CORE_FILTER, "%x", &filter);
+	if (!(0x1 & filter))
+		tst_brk(TCONF, "Anonymous private memory is not dumpable.");
+
+	SAFE_FILE_SCANF(CORE_PATTERN, "%m[^\n]", &cpattern);
+	tst_res(TINFO, "System core pattern is '%s'", cpattern);
+
+	SAFE_GETCWD(cwd, sizeof(cwd));
+	snprintf(tmpcpattern, sizeof(tmpcpattern), "%s/dump-%%p", cwd);
+	tst_res(TINFO, "Temporary core pattern is '%s'", tmpcpattern);
+	SAFE_FILE_PRINTF(CORE_PATTERN, "%s", tmpcpattern);
+
+	fmem = SAFE_MMAP(NULL,
+			 FMEMSIZE,
+			 PROT_READ | PROT_WRITE,
+			 MAP_ANONYMOUS | MAP_PRIVATE,
+			 -1,
+			 0);
+
+	/*
+	 * Write a generated character sequence to the mapped memory,
+	 * which we later look for in the core dump.
+	 */
+	fmemc = (char *)fmem;
+	*fmemc = 'x';
+	for (i = 0; i < YCOUNT; i++)
+		fmemc[i + 1] = 'y';
+	fmemc[++i] = 'z';
+}
+
+static void cleanup(void)
+{
+	if (cpattern) {
+		if (FILE_PRINTF(CORE_PATTERN, "%s", cpattern))
+			tst_res(TWARN,
+				"COULD NOT RESTORE " CORE_PATTERN "! It used to be '%s'",
+				cpattern);
+		free(cpattern);
+	}
+
+	if (fmem && munmap(fmem, FMEMSIZE))
+		tst_res(TWARN | TERRNO, "munmap(fmem)");
+
+	if (dfd > 0 && close(dfd))
+		tst_res(TWARN | TERRNO, "close(dfd)");
+}
+
+static int find_sequence(int pid)
+{
+	char expectc = 'x';
+	ssize_t read, pos = 0;
+	char rbuf[1024];
+	int ycount = 0;
+	char dumpname[256];
+
+	snprintf(dumpname, 256, "dump-%d", pid);
+	tst_res(TINFO, "Dump file should be %s", dumpname);
+	if (access(dumpname, F_OK))
+		tst_brk(TBROK | TERRNO, "Dump file was not found.");
+
+	dfd = SAFE_OPEN(dumpname, O_RDONLY);
+
+	read = SAFE_READ(0, dfd, &rbuf, sizeof(rbuf));
+	while (read) {
+		switch (rbuf[pos]) {
+		case 'x':
+			ycount = 0;
+			expectc = 'y';
+			break;
+		case 'y':
+			if (expectc == 'y') {
+				ycount++;
+			} else {
+				expectc = 'x';
+				break;
+			}
+
+			if (ycount == YCOUNT)
+				expectc = 'z';
+			break;
+		case 'z':
+			if (expectc == 'z')
+				return 1;
+		default:
+			expectc = 'x';
+		}
+		if (++pos >= read) {
+			read = SAFE_READ(0, dfd, &rbuf, sizeof(rbuf));
+			pos = 0;
+		}
+	}
+
+	SAFE_CLOSE(dfd);
+	return 0;
+}
+
+static pid_t run_child(int advice)
+{
+	int status;
+	pid_t pid;
+	char *advstr =
+		advice == MADV_DONTDUMP ? "MADV_DONTDUMP" : "MADV_DODUMP";
+
+	pid = SAFE_FORK();
+	if (pid == 0) {
+		if (madvise(fmem, FMEMSIZE, advice) == -1) {
+			tst_res(TFAIL | TERRNO,
+				"madvise(%p, %lu, %s) = -1",
+				fmem,
+				FMEMSIZE,
+				advstr);
+			exit(1);
+		}
+		abort();
+	}
+
+	SAFE_WAITPID(pid, &status, 0);
+	if (WIFSIGNALED(status) && !WCOREDUMP(status)) {
+		tst_res(TCONF,
+			"No core dump produced after abort(), see core(5).");
+		pid = 0;
+	} else if (!WIFSIGNALED(status)) {
+		pid = 0;
+	}
+
```

Comment:
Is there any need to handle a stop signal in case the child is being traced? I think the answer is no, especially as I was able to trace it without any strange behaviour. It would be interesting to know if there is a reason though.